cranelift/codegen/src/isa/x64/inst/mod.rs

```diff
@@ -37,61 +35,71 @@ use regs::{create_reg_universe_systemv, show_ireg_sized};
 
 /// Instructions.  Destinations are on the RIGHT (a la AT&T syntax).
 #[derive(Clone)]
-pub(crate) enum Inst {
+pub enum Inst {
     /// nops of various sizes, including zero
     Nop { len: u8 },
 
-    /// (add sub and or xor mul adc? sbb?) (32 64) (reg addr imm) reg
+    // =====================================
+    // Integer instructions.
+    /// Integer arithmetic/bit-twiddling: (add sub and or xor mul adc? sbb?) (32 64) (reg addr imm) reg
     Alu_RMI_R {
         is_64: bool,
         op: AluRmiROpcode,
         src: RegMemImm,
         dst: Writable<Reg>,
     },
 
-    /// (imm32 imm64) reg.
-    /// Either: movl $imm32, %reg32 or movabsq $imm64, %reg32
+    /// Constant materialization: (imm32 imm64) reg.
+    /// Either: movl $imm32, %reg32 or movabsq $imm64, %reg32.
     Imm_R {
         dst_is_64: bool,
         simm64: u64,
         dst: Writable<Reg>,
     },
 
-    /// mov (64 32) reg reg
+    /// GPR to GPR move: mov (64 32) reg reg.
     Mov_R_R {
         is_64: bool,
         src: Reg,
         dst: Writable<Reg>,
     },
 
-    /// movz (bl bq wl wq lq) addr reg (good for all ZX loads except 64->64).
-    /// Note that the lq variant doesn't really exist since the default
-    /// zero-extend rule makes it unnecessary.  For that case we emit the
-    /// equivalent "movl AM, reg32".
-    MovZX_M_R {
-        extMode: ExtMode,
-        addr: Addr,
+    /// Zero-extended loads, except for 64 bits: movz (bl bq wl wq lq) addr reg.
+    /// Note that the lq variant doesn't really exist since the default zero-extend rule makes it
+    /// unnecessary. For that case we emit the equivalent "movl AM, reg32".
+    MovZX_RM_R {
+        ext_mode: ExtMode,
+        src: RegMem,
         dst: Writable<Reg>,
     },
 
-    /// A plain 64-bit integer load, since MovZX_M_R can't represent that
-    Mov64_M_R { addr: Addr, dst: Writable<Reg> },
+    /// A plain 64-bit integer load, since MovZX_RM_R can't represent that.
+    Mov64_M_R {
+        src: SyntheticAmode,
+        dst: Writable<Reg>,
+    },
 
-    /// movs (bl bq wl wq lq) addr reg (good for all SX loads)
-    MovSX_M_R {
-        extMode: ExtMode,
-        addr: Addr,
+    /// Loads the memory address of addr into dst.
+    LoadEffectiveAddress {
+        addr: SyntheticAmode,
+        dst: Writable<Reg>,
+    },
+
+    /// Sign-extended loads: movs (bl bq wl wq lq) addr reg.
```

Comment:
pico-nit (smaller even than a nano-nit): These are now sign-extended loads *and* moves.